src/meta/app/src/principal/user_grant.rs

```diff
@@ -48,6 +51,174 @@ pub enum OwnershipObject {
+        }
+    }
+
+    /// Parse encoded key and return the OwnershipObject
+    pub(crate) fn parse_key(p: &mut kvapi::KeyParser) -> Result<Self, kvapi::KeyError> {
+        let q = p.next_raw()?;
+        match q {
+            "database-by-id" => {
+                let db_id = p.next_u64()?;
+                Ok(OwnershipObject::Database {
+                    catalog_name: Self::DEFAULT_CATALOG.to_string(),
+                    db_id,
+                })
+            }
+            "database-by-catalog-id" => {
+                let catalog_name = p.next_str()?;
+                let db_id = p.next_u64()?;
+                Ok(OwnershipObject::Database {
+                    catalog_name,
+                    db_id,
+                })
+            }
+            "table-by-id" => {
+                unimplemented!("No db_id encoded, can not rebuild key");
+                #[allow(unreachable_code)]
+                {
+                    let table_id = p.next_u64()?;
+                    Ok(OwnershipObject::Table {
+                        catalog_name: Self::DEFAULT_CATALOG.to_string(),
+                        db_id: 0,
+                        table_id,
+                    })
+                }
+            }
+            "table-by-catalog-id" => {
+                unimplemented!("No db_id encoded, can not rebuild key");
+                #[allow(unreachable_code)]
+                {
+                    let catalog_name = p.next_str()?;
+                    let db_id = p.next_u64()?;
+                    let table_id = p.next_u64()?;
+                    Ok(OwnershipObject::Table {
+                        catalog_name,
+                        db_id,
+                        table_id,
+                    })
+                }
+            }
+            "stage-by-name" => {
+                let name = p.next_str()?;
+                Ok(OwnershipObject::Stage { name })
+            }
+            "udf-by-name" => {
+                let name = p.next_str()?;
+                Ok(OwnershipObject::UDF { name })
+            }
+            _ => Err(kvapi::KeyError::InvalidSegment {
+                i: p.index(),
+                expect: "database-by-id|database-by-catalog-id|table-by-id|table-by-catalog-id|stage-by-name|udf-by-name"
+                    .to_string(),
+                got: q.to_string(),
+            }),
+        }
+        //
+    }
+}
+
+/// The meta-service key of subject to grant a role to.
+///
+/// It could be a tenant's database, a tenant's table etc.
+/// It is in form of `__fd_object_owners/<tenant>/<object>`.
+/// where `<object>` could be:
+/// - `database-by-id/<db_id>`
+/// - `database-by-catalog-id/<catalog>/<db_id>`
+/// - `table-by-id/<table_id>`
+/// - `table-by-catalog-id/<catalog>/<table_id>`
+/// - `stage-by-name/<stage_name>`
+/// - `udf-by-name/<udf_name>`
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct RoleGrantee {
```

Comment:
the name role grantee might not quite precise in this case, if i understand correctly, here is representing a ownership in a tenant, ownership is granted by the following SQL:

```
GRANT OWNERSHIP ON xx.xx TO ROLE role1;
```

we also support the following statement:

```
GRANT ROLE role2 TO user1;
```

`RoleGrantee` is easy to be interpreted as the second one.

how about call it `TenantOwnershipObject` ?